### PR TITLE
Add A-GEM replay strategy

### PR DESCRIPTION
**Pull request by nmcnamee24**
Adds an A-GEM replay strategy under `replay/agem.py`.

This implementation trains the wrapped torch backbone with gradient projection against a reservoir replay buffer. When the current update conflicts with replay gradients, it projects the gradient to reduce forgetting before applying the optimizer step.

### CODE REVIEW

File: src/pyclad/strategies/replay/agem.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+from torch import Tensor, nn
+from torch.utils.data import DataLoader, TensorDataset
+
+from pyclad.models.torch_backbone import TorchBackbone
+from pyclad.strategies.replay.buffers.reservoir import ReservoirBuffer
+from pyclad.strategies.strategy import ConceptAgnosticStrategy, ConceptAwareStrategy, ConceptIncrementalStrategy
+
+
+class AGEMStrategy(ConceptIncrementalStrategy, ConceptAwareStrategy, ConceptAgnosticStrategy):
+    """A-GEM strategy using a reservoir replay buffer."""
```
Comment (Nyderx) [inline src/pyclad/strategies/replay/agem.py]: Let's add docstring for arguments :)

File: src/pyclad/strategies/replay/agem.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        self._batch_size = batch_size
+        self._replay_batch_size = replay_batch_size
+        self._epochs = epochs
+        self._projection_tolerance = projection_tolerance
+        self._device = torch.device(device)
+        self._current_batch_size = batch_size
+        self._task_count = 0
+
+    def learn(self, data: np.ndarray, *args, **kwargs) -> None:
+        del args, kwargs
```
Comment (Nyderx) [inline src/pyclad/strategies/replay/agem.py]: Why do we do `del`?

File: src/pyclad/strategies/replay/agem.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    ) -> None:
+        self._model = model
+        self._buffer = buffer
+        self._batch_size = batch_size
+        self._replay_batch_size = replay_batch_size
+        self._epochs = epochs
+        self._projection_tolerance = projection_tolerance
+        self._device = torch.device(device)
+        self._current_batch_size = batch_size
+        self._task_count = 0
```
Comment (Nyderx) [inline src/pyclad/strategies/replay/agem.py]: I don't think we need to store `task_count` in AGEM?

File: src/pyclad/strategies/replay/agem.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        self._current_batch_size = x.shape[0]
+        return self._model.compute_loss(x)
+
+    def _project_gradient(self, module: nn.Module) -> None:
+        if len(self._buffer) == 0:
+            return
+
+        parameters = [parameter for parameter in module.parameters() if parameter.requires_grad]
+        current_grad = self._capture_gradients(parameters)
+        if not torch.isfinite(current_grad).all():
```
Comment (Nyderx) [inline src/pyclad/strategies/replay/agem.py:65]: What should happen in such case? simply `return` may lead to people running the experiments with AGEM and believing that everything is okay. I would either raise an Exception (if AGEM should not continue in such case) or add `logger.warning`/debug

File: src/pyclad/strategies/replay/agem.py (hunk 2 above)
Comment (Nyderx) [reply src/pyclad/strategies/replay/agem.py]: Same in other places

File: src/pyclad/strategies/replay/agem.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+        module.zero_grad(set_to_none=True)
+        replay_n = self._replay_batch_size or self._current_batch_size
+        replay_x, _, _ = self._buffer.sample(n=replay_n, target_device=self._device)
+        self._model.compute_loss(replay_x).backward()
+        replay_grad = self._capture_gradients(parameters)
+        final_grad = current_grad
+        if torch.isfinite(replay_grad).all():
+            final_grad = self._project_if_needed(current_grad, replay_grad)
+        module.zero_grad(set_to_none=True)
```
Comment (Nyderx) [inline src/pyclad/strategies/replay/agem.py]: `set_to_none` is `True` by default, we could skip it.

**Review by Nyderx — CHANGES_REQUESTED** [review]

**Comment by nmcnamee24** [comment]
Hey! I fixed the issues you had pointed out. Thank you, let me know if there is anything else you need me to fix.

**Review by Nyderx — APPROVED** [review]
Hi Noah! Thank you for adding AGEM :)
